# DS-SF-36 | 03 | `pandas` | Assignment | Starter Code

## `pandas` and Bistro

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `pandas` (as `pd`).

In [3]:
import os

import numpy as np

import pandas as pd

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-03-bistro.csv` dataset.

In [8]:
df = pd.read_csv(os.path.join("..","datasets","dataset-03-bistro.csv"))

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [10]:
type(df)

pandas.core.frame.DataFrame

Answer: pandas.core.frame.DataFrame

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [26]:
df.index.size

244

Answer: 244

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [25]:
df.columns.size


8

Answer: 8

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [49]:
cols = df.columns.tolist()

for c in cols:
    print c

day
time
name
gender
is_smoker
party
check
tip


> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [55]:
print(df.head(2))

      day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66


Answer: 

     day    time      name  gender  is_smoker  party  check   tip
    0Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
    1Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66


> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [56]:
print(df.tail(2))

          day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00


Answer:

           day     time   name    gender  is_smoker  party  check   tip
    242 Saturday  Dinner     Jon    Male      False      2  17.82  1.75
    243 Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

> ### Question 9.  Does the dataset contain any missing values?

In [61]:
sum(df.index.isnull())

0

Answer: No

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [71]:
sum(df.is_smoker >0)

In [ ]:
df.drop(['is_smoker'], axis = 1, inplace = True)

In [72]:
df.head()

day    time      name  gender  party  check   tip
0  Sunday  Dinner  Kimberly  Female      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      3  10.34  1.66
2  Sunday  Dinner     Larry    Male      3  21.01  3.50
3  Sunday  Dinner    Joseph    Male      2  23.68  3.31
4  Sunday  Dinner    Janice  Female      4  24.59  3.61

Answer: There are no smoker data in dataset so it wont paly any role in analysing dataset. We can safely drop this variables from dataset.

> ### Question 11.  For which week days does the dataset has data for?

In [79]:
df.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: ['Sunday', 'Saturday', 'Thursday', 'Friday']

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [80]:
df.day.value_counts()

Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer: 

    Saturday    87
    Sunday      76
    Thursday    62
    Friday      19

> ### Question 13.  How much tip did waiters collect for each week day?

In [85]:
df[['day','tip']].groupby('day').sum()

tip
day             
Friday     51.96
Saturday  260.40
Sunday    247.39
Thursday  171.83

Answer: 

            tip
    day	
    Friday	51.96
    Saturday	260.40
    Sunday	247.39
    Thursday	171.83

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [90]:
df[['day','tip']].groupby('day').mean()

tip
day               
Friday    2.734737
Saturday  2.993103
Sunday    3.255132
Thursday  2.771452

Answer: 

            tip
    day	
    Friday	2.734737
    Saturday	2.993103
    Sunday	3.255132
    Thursday	2.771452

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [164]:
tp=df[['day','check','tip']].groupby('day').sum()..apply(lambda x: x['tip']*100/x['check'])

SyntaxError: invalid syntax (<ipython-input-164-9d6d5c8b80e7>, line 1)

In [154]:
tp['tip_per']=tp['tip']*100/tp['check']

In [157]:
tp['tip_per']

day
Friday      15.944519
Saturday    14.642375
Sunday      15.203791
Thursday    15.673201
Name: tip_per, dtype: float64

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [158]:
ml = df[df.gender =='Male'].name

In [159]:
fml = df[df.gender =='Female'].name

In [160]:
np.intersect1d(ml,fml)

array(['Casey'], dtype=object)

Answer: Yes, Casey

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [168]:
df.name.unique().size

181

Answer: 181

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [210]:
df[df.name=='Kevin'].index.size

4

In [211]:
df[df.name =='Alice'].index.size

2

Answer: Kevin: 28, Alice: 14

> ### Question 19.  Who are the top 3 female and male patrons?

In [245]:
ml=df[df.gender=='Male']
fml=df[df.gender=='Female']

In [280]:
df[df.gender=='Male'].[['name','day']].groupby(['name']).count().sort_values('day',ascending=0).head(3)

SyntaxError: invalid syntax (<ipython-input-280-b9815cc15ecd>, line 1)

In [279]:
fml[['name','day']].groupby(['name']).count().sort_values('day',ascending=0).head(3)

day
name      
Mary     4
Laura    3
Casey    3

Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [ ]:
# TODO

Answer: TODO